File: 5th Semester/INF318/Base Conocimiento Propuestas/Recursion/Ciclos.cpp
#include <iostream>

using namespace std;

/// ITERATIVO
bool primo(int N)
{
    int i=2,lim=N/2;

    if(N==1)return false;

    while(i<=lim){
        if(N%i==0)return false;
        i++;
    }

    return true;
}
/// RECURSIVO
bool primo(int N,int i)
{
    if(i>N/2)return true;
    if(N%i==0)return false;
    return primo(N,i+1);
}

/// ITERATIVO
void mostrarDivisores(int n){
    int i=2;
    while(i<=n)
    {
        if(n%i==0)cout<<i<<endl;
        ++i;
    }
}

/// RECURSIVO
void mostrarDivisores(int n,int i)
{
    if(i>n)return;

    if(n%i==0)
        cout<<i<<endl;

    mostrarDivisores(n,i+1);

}
/// ITERATIVO
void mostrarDivisoresPrimos(int n){
    int i=1;
    while(i<=n)
    {
        if(n%i==0 && primo(i))cout<<i<<endl;
        ++i;
    }
}

/// RECURSIVO
void mostrarDivisoresPrimos(int n,int i=2)
{
    if(i>n)return;

    if(n%i==0 && primo(i))
    {
        cout<<i<<endl;
    }
    mostrarDivisoresPrimos(n,i+1);

}

/// ITERATIVO
void mostrarPrimos(int a, int b){
    for(;a<=b;a++)
    {
        if(primo(a))cout<<a<<endl;
    }
}

/// RECURSIVO
void mostrarPrimosSob(int a,int b)
{
    if(a>b)return;

    if(primo(a))cout<<a<<endl;

    mostrarPrimos(a+1,b);

}



int main()
{
  
    return 0;
}

File: 5th Semester/INF318/Base Conocimiento Propuestas/Recursion/Ciclos.pl
% PRIMO
primo:-
		read(N),
		primo(N).

primo(N):- primo(N,2).

primo(N,I):- I > N // 2.

primo(N,I):- N mod I =\= 0,
		     I1 is I+1,
			 primo(N,I1).
% PRIMO			 
mostrarDivisores:- read(N),
		   divisores(N,2).
		   
divisores(N,I):- I>N,!.
divisores(N,I):-
					N mod I =:= 0,
					write(I),nl,
					I1 is I+1,
					divisores(N,I1).
divisores(N,I):- I1 is I+1,
				 divisores(N,I1).
				 
% MOSTRAR PRIMO		 
mostrarDivisoresPrimos:- read(N),
						  mostrarDivisoresPrimos(N,2).
		   
mostrarDivisoresPrimos(N,I):- I>N,!.
mostrarDivisoresPrimos(N,I):-
					N mod I =:= 0,
					primo(I),
					write(I),nl,
					I1 is I+1,
					mostrarDivisoresPrimos(N,I1).
mostrarDivisoresPrimos(N,I):- I1 is I+1,
				 mostrarDivisoresPrimos(N,I1).
				 
				 
% PRIMOS ENTRE
mostrarPrimos:-	
					read(A),
					read(B),
					mostrarPrimos(A,B).
					
mostrarPrimos(A,B):- A >B,!.
					
mostrarPrimos(A,B):-
					 primo(A),
					 write(A),nl,
					 X is A+1,
					 mostrarPrimos(X,B).
mostrarPrimos(A,B):-X is A+1, 
					mostrarPrimos(X,B).

